### PR TITLE
WIP: Add ODH-ADR-0008: Upstream Testing ADR

### PR DESCRIPTION
**Pull request by gmfrasca**
## Description
Submit ADR to define strategy and terminology for testing ODH components upstream

## How Has This Been Tested?
Documentation

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: ODH-ADR-0008-upstream-testing.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+There are two parts to this ADR, and therefore this section is split in two sub-sections as well: **Nomenclature** and **Strategy**
+
+### Nomenclature
+Before we can address Testing Strategy, we must first define the types/terms used in testing.  Pulling from several external resources (see [References](#references) section below), here is what we propose as our testing nomenclature, as well at the cost (time+effort) for running the given test class:
+
+| Type/Term            | Description                                                                                     | Cost      |
+|----------------------|-------------------------------------------------------------------------------------------------|-----------|
+| Unit Testing         | White-box tests that execute small blocks of code and check for edge cases, error handling, etc | Very Low  |
+| Functional Testing   | White- or Black-box tests that validate entire features against business requirements           | Low       |
+| Integration Testing  | Black-box tests that ensure components integrate well together                                  | Medium    |
```
Comment (ldimaggi) [inline ODH-ADR-0008-upstream-testing.md:48]: It would be a good idea to refer to the integration, end-to-end, and perf/stress tests as a group as "Non-Functional" tests as we may want to add other types of non -functional tests.

File: ODH-ADR-0008-upstream-testing.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+| Type              | Owner  | Development | Code PR | Manifest PR | Nightly | Pre-release |
+|-------------------|--------|-------------|---------|-------------|---------|-------------|
+| Unit Tests        | Dev    |      x      |    x    |      o      |    o    |      o      |
+| Functional Tests  | Dev    |      x      |    x    |      o      |    o    |      o      |
+| Smoke Tests*      | Dev/QE |      x      |    x    |      x      |    x    |      x      |
+| Integration Tests | Dev/QE |      x      |         |      x      |    x    |      x      |
+| End-to-End Test   | QE     |             |         |             |    x    |      x      |
+| Perf/Stress Tests | QE     |             |         |             |         |      x      |
+
+*Key: `x`: Required, `o`: Run, but only generates non-blocking warnings
```
Comment (ldimaggi) [inline ODH-ADR-0008-upstream-testing.md:72]: Why wouldn't we want to run the unit and functional tests for the pre-release milestones?

File: ODH-ADR-0008-upstream-testing.md (hunk 1 above)
Comment (ldimaggi) [inline ODH-ADR-0008-upstream-testing.md:48]: Black-box tests that ensure components integrate well together

How about saying:
Black-box tests that ensure components integrate well together within the project

File: ODH-ADR-0008-upstream-testing.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+After following this strategy, with the bulk of the testing being performed on-demand and/or in a nightly cadence against ODH, we can achieve our goal of migrating testing to upstream.
+
+
+## Open Questions
+
+* Remove the 'End-to-End' Terminology: This is a proposal to remove the term 'end-to-end' from testing nomenclature. Reasoning being that the term introduces confusion as it has different meanings to different personas: Developers see end-to-end meaning a test of a simple single workflow of their component, integrators see it as installation-test-teardown cycle for *ALL* components in ODH, and the term is also loaded/magic in various CI platforms as well.  Instead, rename this to 'Acceptance' Testing.
+
+* Build Testing:  It could prove to be valuable to have another class of testing that involves Build sanity, or in other words, the capability to ensure builds are not broken by a given change.  This may already be happening with nightly builds in the pipeline, but should we attach an official status/procedure in this doc around it?
```
Comment (ldimaggi) [inline ODH-ADR-0008-upstream-testing.md:95]: This is a good point - One of the advantages of performing nightly builds is that the act of building the builds debugs the build process.

File: ODH-ADR-0008-upstream-testing.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+## Alternatives
+
+Alternatively, we can continue with the status quo.  This means the bulk of the testing would reside in a silo for QE which proved to be difficult to maintain and automate.  Because the bulk of the work is all framed within the concept of a RHODS release, migrating testing to upstream will prove very difficult.  Additionally, there is not a set standard for development testing which has shown to allow for less stringent testing practices, and doesn't provide a wide coverage against bug introduction
```
Comment (ldimaggi) [inline ODH-ADR-0008-upstream-testing.md:99]: This is not sustainable.   ;-)

File: ODH-ADR-0008-upstream-testing.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+Another alternative is to run most or all tests for every code and manifest change without context.  The tradeoff with this strategy is that development velocity will greatly suffer, as we have found the slower/more expensive tests create significant hesitance and roadblocks when opening pull requests, and hinder the ability to commit small changes quickly (rather than submitting large, monolithic changesets all at once, due to the expensive nature of CI in this state)
+
+Given these two alternatives, we found the middle ground, as outlined above (run fast/cheap tests in development, and heavy/expensive tests on periodic cadences), to be the most efficient.
+
+## Risks
+
+Not running **EVERY** test class on every PR inherently introduces the risk of introducing more bugs than if code changes just required the fast-running tests for merge.  However, the time and effort cost of doing such is likely to cause significant delays and hesitation in development and review cycles, which is the reasoning for this tradeoff.
```
Comment (ldimaggi) [inline ODH-ADR-0008-upstream-testing.md:107]: Before its demise, RHOSAK was setting up run a selected subset of E2E and Perf tests on every merge. The duration of each set of tests was to be 5 min. That is one more alternative.

File: ODH-ADR-0008-upstream-testing.md (hunk 2 above)
Comment (gmfrasca) [reply ODH-ADR-0008-upstream-testing.md:72]: I think ideally we would want to run them, but not block the release if an errant unit test is failing.  In a perfect world 100% of all unit tests will pass all the time, and it's certainly a goal everyone would love to achieve, but I feel in practice it would introduce a difficult criteria to achieve, and that in turn may lead to hesitancy to introduce more vigorous testing.

Another thing to consider would the difficulty to maintain a centralized framework that is capable of understanding and invoking every component's particular testing framework.  It's probably likely that this will be hard to achieve in the first iteration of implementing actions resulting from this ADR, and would be something we would need to revist when ready.

File: ODH-ADR-0008-upstream-testing.md (hunk 4 above)
Comment (gmfrasca) [reply ODH-ADR-0008-upstream-testing.md:99]: Heh, there *is* always the alternative to do nothing.... just don't mind that its frequently the worst option available :)  But yes, this is here more for context on what our testing ecosystem looks like if we don't make a change.

File: ODH-ADR-0008-upstream-testing.md (hunk 5 above)
Comment (gmfrasca) [reply ODH-ADR-0008-upstream-testing.md:107]: I think the dependency here is the installation time.  Assuming we have a way to quickly provision a testable cluster AND install RHODS within this 5m timeframe, this could be an option, but the current frameworks that are widely used in our non-functional testing (group nomenclature from a comment above), these were taking much longer... along the scale of 30-60m or longer, which was not sustainable for development and debugging.

Such an alternative would require the prerequisite to introduce this quick-install mechanism, so I can add that as an alternative.

File: ODH-ADR-0008-upstream-testing.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+There are two parts to this ADR, and therefore this section is split in two sub-sections as well: **Nomenclature** and **Strategy**
+
+### Nomenclature
+Before we can address Testing Strategy, we must first define the types/terms used in testing.  Pulling from several external resources (see [References](#references) section below), here is what we propose as our testing nomenclature, as well at the cost (time+effort) for running the given test class:
+
+| Type/Term            | Description                                                                                     | Cost      |
+|----------------------|-------------------------------------------------------------------------------------------------|-----------|
+| Unit Testing         | White-box tests that execute small blocks of code and check for edge cases, error handling, etc | Very Low  |
+| Functional Testing   | White- or Black-box tests that validate entire features against business requirements           | Low       |
```
Comment (HumairAK) [inline ODH-ADR-0008-upstream-testing.md:47]: > business requirements 

odh doesn't have a business per se, user requirements maybe?

File: ODH-ADR-0008-upstream-testing.md (hunk 1 above)
Comment (HumairAK) [reply ODH-ADR-0008-upstream-testing.md:48]: > ensure components integrate well together     

can we be more specific, by components we mean odh components? so like dashboard/notebook controller/modelmesh/dsp etc?

File: ODH-ADR-0008-upstream-testing.md (hunk 2 above)
Comment (HumairAK) [inline ODH-ADR-0008-upstream-testing.md:72]:
```suggestion
Key: `x`: Required, `o`: Run, but only generates non-blocking warnings
```

File: ODH-ADR-0008-upstream-testing.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+Development and QE will partner in developing a set of smoke tests.  These will also run quickly and reside within the component repository.  Development can and should run these while developing features to ensure component sanity, but QE should also be able to run them without a deep understanding of the underlying code.  This likely will leverage simple scripts or top-level commands with basic configuration files, and will not involve building or deploying the code.  Runtime: Less than 5 minutes
+
+Development and QE will also collaborate to generate a standardized methodology for running **Integration Tests**.  These are slower-running in nature and therefore should not block Code-centric PRs, but still should run for changes made to deployment artifacts (`odh-manifests`, for example).  This would use the ocp-ci framework that already exists for most components and remain in the CI actions for `odh-manifests`.
```
Comment (HumairAK) [inline ODH-ADR-0008-upstream-testing.md:82]: > This would use the ocp-ci framework 

I think we want to move away from ocp-ci at some point, and also the adr probably should stay clear of suggesting technologies (it is also stated as a non-goal above)

File: ODH-ADR-0008-upstream-testing.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+Performance testing also should be run regularly, but is very expensive and likely not feasible or cost-effective to run nightly.  Therefore, this type of testing should be integrated and  run on a per-release cadence.
+
+After following this strategy, with the bulk of the testing being performed on-demand and/or in a nightly cadence against ODH, we can achieve our goal of migrating testing to upstream.
+
+
+## Open Questions
+
+* Remove the 'End-to-End' Terminology: This is a proposal to remove the term 'end-to-end' from testing nomenclature. Reasoning being that the term introduces confusion as it has different meanings to different personas: Developers see end-to-end meaning a test of a simple single workflow of their component, integrators see it as installation-test-teardown cycle for *ALL* components in ODH, and the term is also loaded/magic in various CI platforms as well.  Instead, rename this to 'Acceptance' Testing.
```
Comment (HumairAK) [inline ODH-ADR-0008-upstream-testing.md:93]: > Instead, rename this to 'Acceptance' Testing.

+1 to avoiding the use of e2e term, but I'm confused here, what are you defining Acceptance testing as?

File: ODH-ADR-0008-upstream-testing.md [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+## Risks
+
+Not running **EVERY** test class on every PR inherently introduces the risk of introducing more bugs than if code changes just required the fast-running tests for merge.  However, the time and effort cost of doing such is likely to cause significant delays and hesitation in development and review cycles, which is the reasoning for this tradeoff.
+
+## Stakeholder Impacts
+
+| Group                         | Key Contacts                  | Date       | Impacted? |
+| ----------------------------- | ----------------------------- | ---------- | --------- |
+| RHODS QE                      | [Jorge Garcia](@jgarciao)     |            | yes       |
```
Comment (HumairAK) [inline ODH-ADR-0008-upstream-testing.md:113]: This seems like it affects all the working groups as well no?
